From da0774bf6fcf1e73f6c8df0fedcfecc67956a5da Mon Sep 17 00:00:00 2001
From: Liang Dai <liang.dai1@citrix.com>
Date: Mon, 30 May 2016 02:23:35 +0000
Subject: [PATCH] CA-209500: pbis systemd log spam on every boot

---
 config/linux/redhat/lwsmd.service.in |    1 -
 1 file changed, 1 deletion(-)

diff --git a/config/linux/redhat/lwsmd.service.in b/config/linux/redhat/lwsmd.service.in
index ea93463..ef23378 100644
--- a/config/linux/redhat/lwsmd.service.in
+++ b/config/linux/redhat/lwsmd.service.in
@@ -4,7 +4,6 @@ After=network.target
 
 [Service]
 Type=forking
-EnvironmentFile=@MK_LIBEXECDIR@/init-base.sh
 ExecStart=@MK_SBINDIR@/lwsmd --start-as-daemon 
 ExecReload=@MK_BINDIR@/lwsm refresh
 ExecStop=@MK_BINDIR@/lwsm shutdown
-- 
1.7.10.4

