From ba9a809d2886970b9b98181aaada318c439778ed Mon Sep 17 00:00:00 2001
From: Huan Xie <huan.xie@citrix.com>
Date: Mon, 19 Jun 2017 04:23:33 +0100
Subject: [PATCH] CA-214745 Change lwsmd.service from 0755 to 0644

Signed-off-by: Huan Xie <huan.xie@citrix.com>
---
 config/linux/redhat/MakeKitBuild | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/config/linux/redhat/MakeKitBuild b/config/linux/redhat/MakeKitBuild
index 9c1b351..38baa4d 100644
--- a/config/linux/redhat/MakeKitBuild
+++ b/config/linux/redhat/MakeKitBuild
@@ -18,7 +18,7 @@ make()
 {
     # Works for many different RedHat products
     mk_stage DEST="${MK_SYSCONFDIR}/redhat/lwsmd" MODE=0755 SOURCE="lwsmd"
-    mk_stage DEST="${MK_SYSCONFDIR}/redhat/lwsmd.service" MODE=0755 SOURCE="lwsmd.service"
+    mk_stage DEST="${MK_SYSCONFDIR}/redhat/lwsmd.service" MODE=0644 SOURCE="lwsmd.service"
 
     #### SELINUX ####
 
-- 
2.1.4

