From 0e3a7286dfb4bc519fb2b598af56aa901203a565 Mon Sep 17 00:00:00 2001
From: rboulton <rboulton@pbisdev.com>
Date: Fri, 13 May 2016 12:26:13 +0100
Subject: [PATCH 1/1] Do not add capath entries for SLD #54456

---
 domainjoin/libdomainjoin/src/djkrb5conf.c | 65 +++++++++++++++++--------------
 1 file changed, 35 insertions(+), 30 deletions(-)

diff --git a/domainjoin/libdomainjoin/src/djkrb5conf.c b/domainjoin/libdomainjoin/src/djkrb5conf.c
index 3ffed90..c778e79 100644
--- a/domainjoin/libdomainjoin/src/djkrb5conf.c
+++ b/domainjoin/libdomainjoin/src/djkrb5conf.c
@@ -1234,49 +1234,54 @@ Krb5JoinDomain(
                     GCE(ceError = AddChildNode(capathsGroup, addNode));
                     addNode = NULL;
                 }
-                length = strlen(pszLongNameUpper);
                 CT_SAFE_FREE_STRING(pszTail);
                 GCE(ceError = getTail(pszLongNameUpper, &pszTail));
-                for( dn = pszLongNameUpper; dn < &pszLongNameUpper[length]; dn++)
+                if (pszTail)
                 {
-                    if( (*dn!= '.') && (&dn[1] != &pszLongNameUpper[length]) ) continue; 
-                    ch = &dn[1];
-                    if( !strcmp(ch,pszforestUpper) || 
-                        !strcmp(ch,pszTail)) 
+                    length = strlen(pszLongNameUpper);
+                    for( dn = pszLongNameUpper; dn < &pszLongNameUpper[length]; dn++)
                     {
-                        break;
-                    }
-                    CT_SAFE_FREE_STRING(pszParentDomain);
-                    GCE(ceError = CTStrdup(ch, &pszParentDomain));
-                    if(strcmp(pszParentDomain,pszforestUpper))
-                    {
-                        GCE(ceError = CreateValueNode(conf, 3, pszJoinDomain, pszParentDomain, &addNode));
-                        GCE(ceError = AddChildNode(capathsGroup, addNode));
-                        addNode = NULL;
+                        if( (*dn!= '.') && (&dn[1] != &pszLongNameUpper[length]) ) continue; 
+                        ch = &dn[1];
+                        if( !strcmp(ch,pszforestUpper) || 
+                            !strcmp(ch,pszTail)) 
+                        {
+                            break;
+                        }
+                        CT_SAFE_FREE_STRING(pszParentDomain);
+                        GCE(ceError = CTStrdup(ch, &pszParentDomain));
+                        if(strcmp(pszParentDomain,pszforestUpper))
+                        {
+                            GCE(ceError = CreateValueNode(conf, 3, pszJoinDomain, pszParentDomain, &addNode));
+                            GCE(ceError = AddChildNode(capathsGroup, addNode));
+                            addNode = NULL;
+                        }
                     }
                 }
                 CT_SAFE_FREE_STRING(pszTail);
                 GCE(ceError = getTail(pszJoinDomain, &pszTail));
                 length = strlen(pszJoinDomain);
-                for( dn = pszJoinDomain; dn < &pszJoinDomain[length]; dn++)
+                if (pszTail)
                 {
-                    if( (*dn!= '.') && (&dn[1] != &pszJoinDomain[length]) ) continue; 
-                    ch = &dn[1];
-                    if( !strcmp(ch,pszforestUpper) || 
-                        !strcmp(ch,pszTail)) 
-                    {
-                        break;
-                    }
-                    CT_SAFE_FREE_STRING(pszParentDomain);
-                    GCE(ceError = CTStrdup(ch, &pszParentDomain));
-                    if(strcmp(pszParentDomain,pszforestUpper))
+                    for( dn = pszJoinDomain; dn < &pszJoinDomain[length]; dn++)
                     {
-                        GCE(ceError = CreateValueNode(conf, 3, pszJoinDomain, pszParentDomain, &addNode));
-                        GCE(ceError = AddChildNode(capathsGroup, addNode));
-                        addNode = NULL;
+                        if( (*dn!= '.') && (&dn[1] != &pszJoinDomain[length]) ) continue; 
+                        ch = &dn[1];
+                        if( !strcmp(ch,pszforestUpper) || 
+                            !strcmp(ch,pszTail)) 
+                        {
+                            break;
+                        }
+                        CT_SAFE_FREE_STRING(pszParentDomain);
+                        GCE(ceError = CTStrdup(ch, &pszParentDomain));
+                        if(strcmp(pszParentDomain,pszforestUpper))
+                        {
+                            GCE(ceError = CreateValueNode(conf, 3, pszJoinDomain, pszParentDomain, &addNode));
+                            GCE(ceError = AddChildNode(capathsGroup, addNode));
+                            addNode = NULL;
+                        }
                     }
                 }
-                
             }
         }
     }
-- 
1.8.3.1

