From 47ac7fcc63dd4b497898f0abe3aa6286f765797b Mon Sep 17 00:00:00 2001
From: Phus Lu <phus.lu@citrix.com>
Date: Fri, 28 Aug 2015 10:33:25 +0100
Subject: [PATCH] CP-12576: Do symbol link for pam_lsass.so

---
 package/open/rpm/open.spec.in | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/package/open/rpm/open.spec.in b/package/open/rpm/open.spec.in
index 9d37df5..88de868 100644
--- a/package/open/rpm/open.spec.in
+++ b/package/open/rpm/open.spec.in
@@ -561,6 +561,13 @@ execute_auxiliary_scripts()
     fi
 }
 
+symlink_pam_lsass()
+{
+    if [ ! -e "/lib64/security/pam_lsass.so" ]; then
+        run ln -s /lib/security/pam_lsass.so /lib64/security/pam_lsass.so
+    fi
+}
+
 postinstall()
 {
     logfile "Package: PowerBroker Identity Services Open postinstall begins (`date`)"
@@ -645,6 +652,8 @@ postinstall()
 
     execute_auxiliary_scripts
 
+    symlink_pam_lsass
+
     logfile "Package: PowerBroker Identity Services Open postinstall finished"
     exit 0
 }
-- 
2.1.4

